### PR TITLE
refactor(storage): upgrade foyer to v0.9

### PR DESCRIPTION
**Pull request by MrCroxx**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Upgrade foyer to v0.9 .

foyer v0.9 refactored the disk cache implementation, provides lower latency with less memory usage.

TODO: Release foyer v0.9 and fulfill the change log and release docs. https://github.com/MrCroxx/foyer/issues/512

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Review by Li0k — APPROVED** [review]
LGTM， thanks for the PR

**Comment by MrCroxx** [comment]
https://buildkite.com/risingwavelabs/pull-request/builds/50060#018f9a54-7aac-4d30-9928-2e221e51d7fa

image

Could madsim export `Handle`? cc @wangrunji0408

**Comment by wangrunji0408** [comment]
> Could madsim export Handle?

Let me add it... Before it is available, you can workaround it with `#[cfg(not(madsim))]`. 🤪

**Comment by MrCroxx** [comment]
THAMK YE, @wangrunji0408 .

File: Cargo.toml
Comment (hzxa21) [inline Cargo.toml:79]: I think we will wait for the foyer new release before merging this PR?

**Review by hzxa21 — APPROVED** [review]
LGTM

File: grafana/risingwave-dev-dashboard.dashboard.py
Comment (MrCroxx) [reply grafana/risingwave-dev-dashboard.dashboard.py:2630]: Some unexported but essential operations to monitor. Like batch leader/follower queue time and queue rotation duration.

FYI https://github.com/MrCroxx/foyer/blob/main/foyer-common/src/metrics.rs#L118-L125

File: Cargo.toml
Comment (MrCroxx) [reply Cargo.toml:79]: Yep. I'm waiting for this PR to be approved. If there is any need to modify foyer, I can fix it and release a new version.

File: grafana/risingwave-dev-dashboard.dashboard.py
Comment (hzxa21) [reply grafana/risingwave-dev-dashboard.dashboard.py:2630]: Can you include it as a metric description in L2631?

**Review by hzxa21 — APPROVED** [review]

**Pull request merged**
